### PR TITLE
WIP implemented api and guide documentation for official extensions

### LINKED ISSUE #275: Extensions documentation should be available again
> Hi, not work this page https://www.yiiframework.com/doc-2.0/yii-swiftmailer-mailer.html Not Found (#404) exception

related to #19 

- [x] add extensions to the offline docs download package

### PR DESCRIPTION
**Pull request by cebe**
fixes #275 

---

bildschirmfoto von 2018-04-01 02-03-31

---

bildschirmfoto von 2018-04-01 02-05-28

---

bildschirmfoto von 2018-04-01 02-06-02

---

open tasks:

- [x] determine versions from extension repo (currently only master is rendered and it assumes 2.0)
- [ ] ~~crosslinking for core framework classes does not work in the guides and API doc.~~ will be handled in https://github.com/yiisoft-contrib/yiiframework.com/issues/329

### CODE REVIEW

**Comment by cebe** [comment]
Overall navigation experience could be improved but that should be done after this PR with a clean concept that fits overall site structure.

**Comment by cebe** [comment]
also blocked by #300 here.

File: commands/ApiController.php [hunk 1]
```diff
@@ -113,8 +115,61 @@ public function actionGenerate($version)
...
+            $this->stdout("Start generating $extensionName API $version JSON Info...\n");
+            $this->template = 'json';
+            $this->actionIndex($source, $target);
+            $this->splitJson($target);
+
+            $this->stdout("Finished $extensionName API $version.\n\n", Console::FG_GREEN);
+        }
+        file_put_contents("$targetPath/api.json", Json::encode($versions));
+
+        return 0;
```
Comment (samdark) [inline commands/ApiController.php]: https://github.com/yiisoft/yii2/blob/master/framework/console/ExitCode.php?

File: commands/GuideController.php [hunk 2]
```diff
@@ -165,6 +166,114 @@ public function actionGenerate($version)
         return 0;
     }
 
+    public function actionExtension($extensionName)
+    {
+        $extension = Extension::find()->where(['name' => $extensionName])->active()->one();
+
+        if ($extension === null) {
+            $this->stderr("Unknown extension $extensionName.\n", Console::FG_RED);
+            return 1;
```
Comment (samdark) [inline commands/GuideController.php]: https://github.com/yiisoft/yii2/blob/master/framework/console/ExitCode.php?

File: commands/GuideController.php [hunk 3]
```diff
@@ -165,6 +166,114 @@ public function actionGenerate($version)
...
+
+        foreach($versions as $version) {
+
+            $sourcePath = Yii::getAlias("@app/data/extensions/$extensionName/$version");
+
+            if (!file_exists($sourcePath)) {
+                passthru('git clone ' . escapeshellarg($extension->github_url) . ' ' . escapeshellarg($sourcePath),$ret);
+                if ($ret != 0) {
+                    $this->stderr("Failed to clone git repo for extension $extensionName.\n", Console::FG_RED);
+                    return 1;
```
Comment (samdark) [inline commands/GuideController.php]: https://github.com/yiisoft/yii2/blob/master/framework/console/ExitCode.php?

File: commands/GuideController.php [hunk 4]
```diff
@@ -165,6 +166,114 @@ public function actionGenerate($version)
...
+
+        // TODO determine versions
+        $versions = ['2.0'];
+
+        foreach($versions as $version) {
+
+            $sourcePath = Yii::getAlias("@app/data/extensions/$extensionName/$version");
+
+            if (!file_exists($sourcePath)) {
+                passthru('git clone ' . escapeshellarg($extension->github_url) . ' ' . escapeshellarg($sourcePath),$ret);
```
Comment (samdark) [inline commands/GuideController.php]: `$ret` → `$return`

File: commands/GuideController.php [hunk 5]
```diff
@@ -165,6 +166,114 @@ public function actionGenerate($version)
...
+                if ($ret != 0) {
+                    $this->stderr("Failed to clone git repo for extension $extensionName.\n", Console::FG_RED);
+                    return 1;
+                }
+                // TODO check out correct branch
+            } else {
+                passthru('git -C ' . escapeshellarg($sourcePath) . ' pull', $ret);
+                if ($ret != 0) {
+                    $this->stderr("Failed to update git repo for extension $extensionName.\n", Console::FG_RED);
+                    return 1;
```
Comment (samdark) [inline commands/GuideController.php]: https://github.com/yiisoft/yii2/blob/master/framework/console/ExitCode.php?

File: commands/GuideController.php [hunk 6]
```diff
@@ -165,6 +166,114 @@ public function actionGenerate($version)
...
+                $this->generateExtensionGuideIndex($source, $target, $version, $language, $extension);
+                $this->stdout("Finished $extensionName guide $version in $language.\n\n", Console::FG_CYAN);
+
+                $metadata[$version][] = $language;
+            }
+
+        }
+        file_put_contents("$targetPath/guide.json", Json::encode($metadata));
+
+        return 0;
```
Comment (samdark) [inline commands/GuideController.php]: https://github.com/yiisoft/yii2/blob/master/framework/console/ExitCode.php?

File: jobs/ExtensionImportJob.php [hunk 7]
```diff
@@ -31,7 +31,25 @@ public function execute($queue)
         // do not update timestamps and blame on automated updates
         $extension->detachBehavior('blameable');
         $extension->detachBehavior('timestamp');
+        $extension->detachBehavior('timestamp');
         $extension->save(false);
+
+        if ($extension->isOfficialExtension) {
+            echo "updating extension {$extension->name} docs...\n";
+            passthru(\Yii::getAlias('@app/yii') . ' guide/extension ' . escapeshellarg($extension->name) . ' --interactive=0', $ret);
+            if ($ret != 0) {
```
Comment (samdark) [inline jobs/ExtensionImportJob.php]: $ret → $return?

File: commands/ApiController.php (hunk 1 above)
Comment (cebe) [reply commands/ApiController.php]: this code base is older than that class :) but I'm happy to add it.

**Pull request merged**